examples/nlp/language_modeling/megatron_change_num_partitions.py

```diff
@@ -171,67 +499,290 @@ def main():
+                elif pp_size > 1:
+                    logging.warning(
+                        f"Model config has `pipeline_model_parallel_split_rank` set to "
+                        f"{model.cfg.pipeline_model_parallel_split_rank} and target PP "
+                        f"size is {tgt_pp_size}. "
+                        f"Provided `pipeline_model_parallel_split_rank` is "
+                        f"{pipeline_model_parallel_split_rank}. "
+                        f"Be careful that the model config is correct "
+                        f"if encoder-decoder models are being converted."
+                    )
 
-    if tgt_tp_size > 1:
-        partitions = []
-        params = [p for _, p in model.named_parameters()]
-        partitions.append(params)
+            model.cfg.global_batch_size = old_global_batch_size  # Used for restoration
 
-        model.cfg.tensor_model_parallel_size = tgt_tp_size
-        app_state.model_parallel_size = tgt_tp_size
-        trainer = Trainer(devices=1, strategy=NLPDDPStrategy(), accelerator="cpu", precision=precision)
-        if args.tokenizer_model_path is not None:
-            model.cfg.tokenizer.model = args.tokenizer_model_path
-        model = cls(model.cfg, trainer).to('cpu')
-        model._save_restore_connector = NLPSaveRestoreConnector()
-        split_partition(model, partitions, tgt_tp_size, args.target_file, args.megatron_legacy)
+            # Override flag that forces Model to use AppState instead of Trainer
+            # to determine the world size, global and local rank
+            # Used for simulating load of a specific rank on a single gpu
+            os.environ[NEMO_MEGATRON_MODEL_PARALLEL_APPSTATE_OVERRIDE] = "true"
+
+            # Compute the global rank
+            global_rank = (
+                pp_rank * tgt_tp_size + 0
+            )  # tp_rank = 0 needed just for modules, all TP will be merged to this PP rank
+
+            # Update AppState
+            app_state.world_size = world_size
+            app_state.global_rank = global_rank
+            app_state.local_rank = global_rank % num_gpu_per_node
+            app_state.pipeline_model_parallel_size = tgt_pp_size
+            app_state.tensor_model_parallel_size = tgt_tp_size
+            app_state.model_parallel_size = (
+                app_state.pipeline_model_parallel_size * app_state.tensor_model_parallel_size
+            )
+
+            trainer = Trainer(devices=1, strategy=NLPDDPStrategy(), accelerator="cpu", precision=precision)
+            if args.tokenizer_model_path is not None:
+                model.cfg.tokenizer.model = args.tokenizer_model_path
+
+            model = cls(model.cfg, trainer).to('cpu')
+            model._save_restore_connector = NLPSaveRestoreConnector()
+            model.to(dtype=dtype)
+
+            # Update global batch size
+            if old_global_batch_size % new_global_batch_size != 0 or old_global_batch_size < new_global_batch_size:
```

Comment:
nit: can we just keep the first condition here? I don't think we need the second one

Reply:
The second one is needed because just being divisible is not enough, at minimum it must be the size of the new world size or assert breaks and model can't be built

Reply:
just to double-check, did you intend the second condition to be `old_global_batch_size < new_global_batch_size`?
if so, condition 1 is always True when condition 2 is since GBS is >0

but its just a nit, don't mind keeping it the same way if you prefer that

Reply:
I don't remember exactly what config of tp pp conversion I needed both flags for, but it is needed for at least one weirs T5 config conversion

Reply:
sounds good